packages/@expo/cli/src/utils/codesigning.ts

```diff
@@ -0,0 +1,350 @@
+ * This requires the user be logged in and online, otherwise try to use the cached development certificate.
+ */
+async function getExpoRootDevelopmentCodeSigningInfoAsync(
+  exp: ExpoConfig
+): Promise<CodeSigningInfo | null> {
+  const easProjectId = exp.extra?.eas?.projectId;
+  // can't check for scope key validity since scope key is derived on the server from projectId and we may be offline.
+  // we rely upon the client certificate check to validate the scope key
+  if (!easProjectId) {
+    Log.warn('No project ID specified in app.json, unable to sign manifest');
+    return null;
+  }
+
+  // 1. If online, ensure logged in, generate key pair and CSR, fetch and cache certificate chain for projectId
+  //    (overwriting existing dev cert in case projectId changed or it has expired)
+  if (!APISettings.isOffline) {
+    return await fetchAndCacheNewDevelopmentCodeSigningInfoAsync(easProjectId);
+  }
+
+  // 2. check for cached cert/private key matching projectId and scopeKey of project, if found and valid return private key and cert chain including expo-go cert
+  const developmentCodeSigningInfoFromFile = await DevelopmentCodeSigningInfoFile.readAsync(
+    easProjectId
+  );
+  const validatedCodeSigningInfo = validateStoredDevelopmentExpoRootCertificateCodeSigningInfo(
+    developmentCodeSigningInfoFromFile,
+    easProjectId
+  );
+  if (validatedCodeSigningInfo) {
+    return validatedCodeSigningInfo;
+  }
+
+  // 3. if offline, return null
+  Log.warn('Offline and no cached development certificate found, unable to sign manifest');
+  return null;
+}
+
+/**
+ * Get the certificate configured for expo-updates for this project.
+ */
+async function getProjectCodeSigningCertificateAsync(
+  exp: ExpoConfig,
+  privateKeyPath: string | undefined,
+  expectedKeyId: string,
+  expectedAlg: string | null
+): Promise<CodeSigningInfo | null> {
+  const codeSigningCertificatePath = exp.updates?.codeSigningCertificate;
+  if (!codeSigningCertificatePath) {
+    return null;
+  }
+
+  if (!privateKeyPath) {
+    privateKeyPath = path.join(path.dirname(codeSigningCertificatePath), 'private-key.pem');
+  }
+
+  const codeSigningMetadata = exp.updates?.codeSigningMetadata;
+  if (!codeSigningMetadata) {
+    throw new Error(
+      'Must specify "codeSigningMetadata" under the "updates" field of your app config file to use EAS code signing'
+    );
+  }
+
+  const { alg, keyid } = codeSigningMetadata;
+  if (!alg || !keyid) {
+    throw new Error(
+      'Must specify "keyid" and "alg" in the "codeSigningMetadata" field under the "updates" field of your app config file to use EAS code signing'
+    );
+  }
+
+  if (expectedKeyId !== keyid) {
+    throw new Error(`keyid mismatch: client=${expectedKeyId}, project=${keyid}`);
+  }
+
+  if (expectedAlg && expectedAlg !== alg) {
+    throw new Error(`alg mismatch: client=${expectedAlg}, project=${alg}`);
```

Comment:
```suggestion
    throw new CommandError(`algorithm mismatch (client: ${expectedAlg}, project: ${alg})`);
```
- bit more descriptive.
- CommandError
- I've been trying to stick to the `(foo: bar, baz: foz)` format from EAS CLI.

Reply:
"alg" is written this way because there is literally a field named "alg" but maybe it is helpful to write it as
```
`"alg" field mismatch: ...`
```